File: bmc-simulator/src/main/kotlin/com/odim/simulator/http/Response.kt
/*
 * Copyright (c) Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.odim.simulator.http

import com.fasterxml.jackson.databind.node.ObjectNode
import com.odim.simulator.behaviors.ExpandType
import com.odim.simulator.behaviors.ExpandType.ALL
import com.odim.simulator.behaviors.TreeJsonRenderer
import com.odim.simulator.dsl.makeJson
import com.odim.simulator.http.HttpStatusCode.ACCEPTED
import com.odim.simulator.http.HttpStatusCode.BAD_REQUEST
import com.odim.simulator.http.HttpStatusCode.CREATED
import com.odim.simulator.http.HttpStatusCode.INTERNAL_SERVER_ERROR
import com.odim.simulator.http.HttpStatusCode.METHOD_NOT_ALLOWED
import com.odim.simulator.http.HttpStatusCode.NOT_FOUND
import com.odim.simulator.http.HttpStatusCode.NOT_MODIFIED
import com.odim.simulator.http.HttpStatusCode.NO_CONTENT
import com.odim.simulator.http.HttpStatusCode.OK
import com.odim.simulator.http.HttpStatusCode.PRECONDITION_FAILED
import com.odim.simulator.http.HttpStatusCode.UNAUTHORIZED
import com.odim.simulator.tree.structure.Item
import com.odim.simulator.tree.structure.Resource
import com.odim.utils.JsonMapper.emptyJson
import com.odim.utils.toJson
import java.util.Collections.singletonList

const val LOCATION_HEADER_NAME = "Location"

data class Response constructor(
    val code: HttpStatusCode,
    val body: String,
    val headers: MutableMap<String, List<String>> = mutableMapOf()
) {
    val json: ObjectNode by lazy {
        if (body.isBlank()) emptyJson else body.toJson()
    }
    val location = headers[LOCATION_HEADER_NAME].orEmpty().firstOrNull()

    companion object {
        fun created(resource: Resource, extraHeaders: Map<String, List<String>> = mapOf()): Response {
            val headers = mutableMapOf(LOCATION_HEADER_NAME to singletonList(resource.toLink()))
            headers.putAll(extraHeaders)
            return Response(CREATED, TreeJsonRenderer().toJson(resource), headers)
        }
        fun success(
            item: Item,
            headers: MutableMap<String, List<String>> = mutableMapOf(),
            expandType: ExpandType = ALL,
            expandLevels: Int = 0
        ) = Response(OK, TreeJsonRenderer(expandType, expandLevels).toJson(item), headers)
        fun success(headers: MutableMap<String, List<String>> = mutableMapOf()) = Response(OK, "", headers)
        fun accepted(location: String, body: String) = Response(ACCEPTED, body, mutableMapOf(LOCATION_HEADER_NAME to listOf(location)))
        fun unauthorized() = predefinedResponse(UNAUTHORIZED)
        fun notAllowed() = predefinedResponse(METHOD_NOT_ALLOWED)
        fun notFound() = predefinedResponse(NOT_FOUND)
        fun noContent() = Response(NO_CONTENT, "")
        fun internalServerError() = Response(INTERNAL_SERVER_ERROR,
                errorJson("The request failed due to an internal service error.  The service is still operational."))
        fun badRequest(message: String = "Bad request") = Response(BAD_REQUEST, errorJson(message, "InvalidPayload"))
        fun notModified() = predefinedResponse(NOT_MODIFIED)
        fun preconditionFailed() = predefinedResponse(PRECONDITION_FAILED)
        fun errorJson(message: String, errorType: String = "GeneralError") = makeJson {
            "error" to makeJson {
                "code" to "Base.1.0.$errorType"
                "message" to message
            }
        }.toString()

        private fun predefinedResponse(httpStatusCode: HttpStatusCode) = Response(httpStatusCode, "")
    }
}

File: lib-utilities/response/response.go
//(C) Copyright [2020] Hewlett Packard Enterprise Development LP
//
//Licensed under the Apache License, Version 2.0 (the "License"); you may
//not use this file except in compliance with the License. You may obtain
//a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
//WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
//License for the specific language governing permissions and limitations
// under the License.

package response

const (
	// BaseVersion defines the latest version of Base
	BaseVersion = "Base.1.11.0."
	// Created is the message for successful creation
	Created = BaseVersion + "Created"
	// ExtendedInfo message
	ExtendedInfo = BaseVersion + "ExtendedInfo"
	// AccountRemoved is the message for successful removal of account
	AccountRemoved = BaseVersion + "AccountRemoved"
	// Success is the message for successful completion
	Success = BaseVersion + "Success"
	// AccountModified is the message for successful account modification
	AccountModified = BaseVersion + "AccountModified"
	// GeneralError defines the code at the time of General Error
	GeneralError = BaseVersion + "GeneralError"
	// Failure code for failed message
	Failure = BaseVersion + "Failed"
	// InsufficientPrivilege defines the status message at the time of Insufficient Privileges
	InsufficientPrivilege = BaseVersion + "InsufficientPrivilege"
	// InternalError defines the status message at the time of Internal Error
	InternalError = BaseVersion + "InternalError"
	// PropertyMissing defines the status message at the time of Property Missing
	PropertyMissing = BaseVersion + "PropertyMissing"
	// PropertyUnknown defines the status message at the time of Property Unknown
	PropertyUnknown = BaseVersion + "PropertyUnknown"
	// ResourceNotFound defines the status message at the time of Resource Not Found
	ResourceNotFound = BaseVersion + "ResourceNotFound"
	// MalformedJSON defines the status message at the time of Malformed JSON
	MalformedJSON = BaseVersion + "MalformedJSON"
	// PropertyValueNotInList defines the status message at the time of Property Value Not In List
	PropertyValueNotInList = BaseVersion + "PropertyValueNotInList"
	// NoValidSession defines the status message at the time of No Valid Session
	NoValidSession = BaseVersion + "NoValidSession"
	// ResourceInUse defines events aleady subscribed
	ResourceInUse = BaseVersion + "ResourceInUse"
	// PropertyValueFormatError defines the status message  given the correct value type but the value of that property was not supported
	PropertyValueFormatError = BaseVersion + "PropertyValueFormatError"
	// PropertyValueTypeError defines the message that the property is value given is having a different format
	PropertyValueTypeError = BaseVersion + "PropertyValueTypeError"
	// ResourceAtURIUnauthorized defines the authorization failure with plugin or other resources
	ResourceAtURIUnauthorized = BaseVersion + "ResourceAtUriUnauthorized"
	// CouldNotEstablishConnection defines the connection failure with plugin or other resources
	CouldNotEstablishConnection = BaseVersion + "CouldNotEstablishConnection"
	// QueryCombinationInvalid defines the status message at the time of invalid query
	QueryCombinationInvalid = BaseVersion + "QueryCombinationInvalid"
	// QueryNotSupported defines the status message at the time of not supported query
	QueryNotSupported = BaseVersion + "QueryNotSupported"
	// ResourceRemoved is the message for successful removal of resource
	ResourceRemoved = "ResourceEvent.1.0.3.ResourceRemoved"
	// ResourceCreated is the message for successful creation of resource
	ResourceCreated = "ResourceEvent.1.0.3.ResourceCreated"
	// TaskStarted is the message for denoting the starting of the task
	TaskStarted = "TaskEvent.1.0.3.TaskStarted"
	// ActionNotSupported defines requested POST operation is not supported by the resource
	ActionNotSupported = BaseVersion + "ActionNotSupported"
	// ResourceAlreadyExists indicates the request is for creation of a resource, which already exists
	ResourceAlreadyExists = BaseVersion + "ResourceAlreadyExists"
	// ActionParameterNotSupported indicates that the parameter supplied for the action is not supported on the resource.
	ActionParameterNotSupported = BaseVersion + "ActionParameterNotSupported"
	// ResourceCannotBeDeleted indicates the requested delete operation cannot be performed
	ResourceCannotBeDeleted = BaseVersion + "ResourceCannotBeDeleted"
	// PropertyValueConflict indicates that the requested write of a property value could not be completed, because of a conflict with another property value.
	PropertyValueConflict = BaseVersion + "PropertyValueConflict"
	// NoOperation  defines the status message at the time of of there is no opeartion need to be performed.
	NoOperation = BaseVersion + "NoOperation"
)

// Response holds the generic response from odimra
type Response struct {
	OdataType    string   `json:"@odata.type"`
	OdataID      string   `json:"@odata.id"`
	OdataContext string   `json:"@odata.context,omitempty"`
	Description  string   `json:"Description,omitempty"`
	ID           string   `json:"Id,omitempty"`
	Name         string   `json:"Name"`
	Message      string   `json:"Message,omitempty"`
	MessageID    string   `json:"MessageId,omitempty"`
	MessageArgs  []string `json:"MessageArgs,omitempty"`
	NumberOfArgs int      `json:"NumberOfArgs,omitempty"`
	Severity     string   `json:"Severity,omitempty"`
	Resolution   string   `json:"Resolution,omitempty"`
}

// RPC defines the reponse which odimra service returns back as
// part of the RPC call.
//
// StatusCode defines the status code of the requested service operation.
// StatusMessage defines the message regarding the status of the requested operation.
// Header defines the headers required to create a proper response from the api gateway.
// Body defines the actual response of the requested service operation.
type RPC struct {
	StatusCode    int32
	StatusMessage string
	Header        map[string]string
	Body          interface{}
}

//CommonError holds the error response from odimra
type CommonError struct {
	Error ErrorClass `json:"error"`
}

//ErrorClass holds the properties that describe error from odimra
//
//Code indicates a specific MessageId from a Message Registry.
//Message contains error message corresponding to the message in a Message Registry.
//MessageExtendedInfo is an message objects describing one or more error messages.
type ErrorClass struct {
	Code                string `json:"code"`
	Message             string `json:"message"`
	MessageExtendedInfo []Msg  `json:"@Message.ExtendedInfo,omitempty"`
}

// Msg holds the properties of Message object
type Msg struct {
	OdataType   string        `json:"@odata.type,omitempty"`
	MessageID   string        `json:"MessageId,omitempty"`
	Message     string        `json:"Message,omitempty"`
	Severity    string        `json:"Severity,omitempty"`
	MessageArgs []interface{} `json:"MessageArgs,omitempty"`
	Resolution  string        `json:"Resolution,omitempty"`
}

//Args holds the slice of ErrArgs
type Args struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	ErrorArgs []ErrArgs
}

// ErrArgs holds the parameters to build error response
type ErrArgs struct {
	StatusMessage string
	ErrorMessage  string
	MessageArgs   []interface{}
}
